### PR TITLE
report gpu model and driver version

### PR DESCRIPTION
**Pull request by buptcozy**
Signed-off-by: yangzhang <bupt_cozy@126.com>

### V. Checklist

- [ ] I have written necessary docs and comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `make test`

### CODE REVIEW

File: pkg/slo-controller/noderesource/noderesource.go [hunk 1]
```diff
@@ -195,6 +199,88 @@ func (r *NodeResourceReconciler) updateGPUNodeResource(node *corev1.Node, device
...
+			continue
+		}
+		hasGPUDevice = true
+	}
+
+	if !hasGPUDevice {
+		return nil
+	}
+
+	modelList, driverVersion, err := getGPUModelAndDriverInfo()
```
Comment (zwzhang0107) [inline pkg/slo-controller/noderesource/noderesource.go:219]: maybe this should be implemented in koordlet?

File: go.mod [hunk 2]
```diff
@@ -4,6 +4,7 @@ go 1.17
 
 require (
 	github.com/NVIDIA/go-nvml v0.11.6-0.0.20220823120812-7e2082095e82
+	github.com/NVIDIA/gpu-monitoring-tools v0.0.0-20211102125545-5a2c58442e48
```
Comment (ZYecho) [inline go.mod:7]: Deprecated?

File: pkg/slo-controller/noderesource/noderesource.go (hunk 1 above)
Comment (ZYecho) [reply pkg/slo-controller/noderesource/noderesource.go:219]: +1

File: apis/extension/resource.go [hunk 3]
```diff
@@ -42,6 +42,9 @@ const (
 	GPUCore        corev1.ResourceName = ResourceDomainPrefix + "gpu-core"
 	GPUMemory      corev1.ResourceName = ResourceDomainPrefix + "gpu-memory"
 	GPUMemoryRatio corev1.ResourceName = ResourceDomainPrefix + "gpu-memory-ratio"
+
+	GpuDriver string = ResourceDomainPrefix + "gpu-driver"
```
Comment (eahydra) [inline apis/extension/resource.go:46]: Keep GPU prefixes consistent with the style of existing variables.

File: go.mod (hunk 2 above)
Comment (eahydra) [reply go.mod:7]: +1.  And we already depend on github.com/NVIDIA/go-nvml

File: pkg/slo-controller/noderesource/noderesource.go [hunk 4]
```diff
@@ -195,6 +199,88 @@ func (r *NodeResourceReconciler) updateGPUNodeResource(node *corev1.Node, device
...
+	if err != nil {
+		klog.Errorf("failed to get GPU Model and driver info: %v", err)
```
Comment (eahydra) [inline pkg/slo-controller/noderesource/noderesource.go:221]: should add the  name of node name or device CRD object in the log.

File: pkg/slo-controller/noderesource/noderesource.go [hunk 5]
```diff
@@ -195,6 +199,88 @@ func (r *NodeResourceReconciler) updateGPUNodeResource(node *corev1.Node, device
...
+	if err = r.Client.Patch(context.Background(), copyNode, patch); err != nil {
+		klog.Errorf("patch node gpu model and version fail, err:%v", err)
+		return err
+	}
+	return nil
+}
+
+func getGPUModelAndDriverInfo() ([]*string, string, error) {
+	var modelList []*string
+	driverVersion, err := nvmlbindings.GetDriverVersion()
```
Comment (eahydra) [inline pkg/slo-controller/noderesource/noderesource.go:266]: hmm.. Does it work?  slo-controller just a add-on controller works in the cluster not the node.

File: pkg/slo-controller/noderesource/noderesource.go [hunk 6]
```diff
@@ -195,6 +199,88 @@ func (r *NodeResourceReconciler) updateGPUNodeResource(node *corev1.Node, device
 	return err
 }
 
+func (r *NodeResourceReconciler) updateGPUDriverAndModel(node *corev1.Node, device *schedulingv1alpha1.Device) error {
```
Comment (eahydra) [inline pkg/slo-controller/noderesource/noderesource.go:202]: please add the UT.

File: pkg/slo-controller/noderesource/noderesource.go (hunk 1 above)
Comment (eahydra) [reply pkg/slo-controller/noderesource/noderesource.go:219]: The key question is whether we allow koordlet to modify Node. It seems that there are no features of koordlet to directly modify Node in the current code?

**Review by eahydra — CHANGES_REQUESTED** [review]

File: pkg/slo-controller/noderesource/noderesource.go (hunk 1 above)
Comment (buptcozy) [reply pkg/slo-controller/noderesource/noderesource.go]: fixed

File: apis/extension/resource.go (hunk 3 above)
Comment (buptcozy) [reply apis/extension/resource.go]: fixed

File: go.mod (hunk 2 above)
Comment (buptcozy) [reply go.mod:7]: fixed

File: pkg/slo-controller/noderesource/noderesource.go (hunk 4 above)
Comment (buptcozy) [reply pkg/slo-controller/noderesource/noderesource.go]: fixed

File: pkg/slo-controller/noderesource/noderesource.go (hunk 5 above)
Comment (buptcozy) [reply pkg/slo-controller/noderesource/noderesource.go]: fixed

File: pkg/slo-controller/noderesource/noderesource.go (hunk 6 above)
Comment (buptcozy) [reply pkg/slo-controller/noderesource/noderesource.go:199]: fixed

File: pkg/slo-controller/noderesource/noderesource.go [hunk 7]
```diff
@@ -195,6 +196,30 @@ func (r *NodeResourceReconciler) updateGPUNodeResource(node *corev1.Node, device
...
+	if device == nil || device.Labels == nil {
+		return nil
+	}
+
+	if device.Labels[extension.GPUModel] == "" && device.Labels[extension.GPUDriver] == "" {
```
Comment (ZYecho) [inline pkg/slo-controller/noderesource/noderesource.go:204]: 这个地方有个问题，device是个通用对象，包含不同的device类型，包括gpu/fpga等，不同device可能有差异化的信息，这部分差异化的信息，我们目前是通过label来设置？

那么随着不同device的增加（差异化的字段也不断增加），这个函数要一直增加函数入参么？
s.createDevice(node, gpuDevices, gpuModel, gpuDriverVer)

File: pkg/slo-controller/noderesource/noderesource.go [hunk 8]
```diff
@@ -195,6 +196,30 @@ func (r *NodeResourceReconciler) updateGPUNodeResource(node *corev1.Node, device
...
+	copyNode := node.DeepCopy()
+	if copyNode.Labels == nil {
+		copyNode.Labels = make(map[string]string)
+	}
+	copyNode.Labels[extension.GPUModel] = device.Labels[extension.GPUModel]
+	copyNode.Labels[extension.GPUDriver] = device.Labels[extension.GPUDriver]
+
+	patch := client.MergeFrom(node)
+	if err := r.Client.Patch(context.Background(), copyNode, patch); err != nil {
+		klog.Errorf("patch node gpu model and version fail, err:%v", err)
```
Comment (ZYecho) [inline pkg/slo-controller/noderesource/noderesource.go:217]: maybe

failed to patch node gpu model and version, err:%v

is better?

File: pkg/koordlet/statesinformer/states_device_linux.go [hunk 9]
```diff
@@ -119,6 +131,17 @@ func (s *statesInformer) updateDevice(name string, gpuDevices []schedulingv1alph
 		device.Spec = schedulingv1alpha1.DeviceSpec{
 			Devices: gpuDevices,
 		}
+
+		if device.Labels == nil {
```
Comment (eahydra) [inline pkg/koordlet/statesinformer/states_device_linux.go:135]: It is necessary to pay attention to the existing scenes. If device.Spec.Devices does not change, the following label will not be updated.

**Review by eahydra — CHANGES_REQUESTED** [review]
Please resolve the errors of Github Actions

File: pkg/koordlet/statesinformer/states_device_linux.go (hunk 9 above)
Comment (zwzhang0107) [reply pkg/koordlet/statesinformer/states_device_linux.go:135]: +1，the DeepEqual on L126

File: pkg/slo-controller/noderesource/noderesource.go (hunk 7 above)
Comment (buptcozy) [reply pkg/slo-controller/noderesource/noderesource.go:204]: fixed

File: pkg/koordlet/statesinformer/states_device_linux.go (hunk 9 above)
Comment (buptcozy) [reply pkg/koordlet/statesinformer/states_device_linux.go]: fixed

File: pkg/slo-controller/noderesource/noderesource.go (hunk 8 above)
Comment (buptcozy) [reply pkg/slo-controller/noderesource/noderesource.go]: fixed

**Pull request closed without merging**